Smt/String.lean

```diff
@@ -10,24 +10,45 @@ import Smt.Transformer
 
 namespace Smt.String
 
-open Lean
-open Lean.Expr
+open Lean Expr
 open Smt.Transformer
 
-@[Smt] def transformStringGetOp : Transformer
+@[Smt] def replaceConst : Transformer
+  | const `Char.ofNat ..     => pure none
+  | const `String.Pos.mk ..  => pure none
+  | const `String.replace .. => pure (mkConst `str.replace_all)
+  | const `String.length ..  => pure (mkConst `str.len)
+  | const `String.append ..  => pure (mkConst (Name.mkSimple "str.++"))
+  | e                        => pure e
+
+@[Smt] def removeStringPosMk : Transformer
+  | app (const `String.Pos.mk ..) e _ => applyTransformations e
+  | e                                 => pure e
+
+@[Smt] def replaceStringGetOp : Transformer
   | app (app (const `String.getOp ..) f _) e _ => do
     return match ← applyTransformations f, ← applyTransformations e with
     | some f', some e' =>
       mkApp (mkConst `str.to_code) (mkApp2 (mkConst `str.at) f' e')
     | _      , _       => none
-  | e                                          => return e
+  | e                                          => pure e
 
-@[Smt] def transformStringContains : Transformer
+@[Smt] def replaceStringContains : Transformer
   | app (app (const `String.contains ..) f _) e _ => do
     return match ← applyTransformations f, ← applyTransformations e with
     | some f', some e' =>
       mkApp2 (mkConst `str.contains) f' (mkApp (mkConst `str.from_code) e')
     | _      , _       => none
-  | e                                             => return e
+  | e                                             => pure e
+
+@[Smt] def replaceStringLt : Transformer
```

Comment:
It's a bit unfortunate that unfolding the type class results in having to handle this low-level expression.

Reply:
So to think about it more, *why* is the type class unfolding pass needed? It looks like it will result in lots of special cases dealing with how instances at particular types are defined, such as this one. I agree that the generic "type-removal" pass is good to get rid of, but maybe it would be better to keep the class projections and just handle them by `@Eq Nat => '='`, `@Eq Bool => '='`, or more generally `@Eq T => '='` for every type `T` where we have equality defined in SMTLIB. In other words, a type removal pass but specialized to known predicates at known types, without unfolding.

Reply:
Okay, I tried it but it does not really improve things with the current setup because some transformers rely on the existence of primitive `Nat` operations. Let's merge this PR and discuss separately.

Reply:
That's a good point you're raising. I considered this option for a while because unfolding is not the kind of silver bullet I hoped it would be. It does not get rid of all type class instantiations and it sometimes unfolds more things than necessary. This particular case is worrying because I plan on supporting Lists in the future. But overall, I think it does reduce the number of corner cases I have to deal with. Here are 3 reasons:
- Supporting `0 + 1 = 1` does not mean it's ok to reject `Nat.add 0 1 = 1`.
- We still have projections for user-defined instances. This is not common for builtin types, but it's common for user-defined ones.
- It does an amazing job with literals. This is how `0 + 1 = 1` looks before unfolding:
```
(APP (CONST Not []) (APP (APP (APP (CONST Eq [1]) (CONST Nat [])) (APP (APP (APP (APP (APP (APP (CONST HAdd.hAdd [0, 0, 0]) (CONST Nat [])) (CONST Nat [])) (CONST Nat [])) (APP (APP (CONST instHAdd [0]) (CONST Nat [])) (CONST instAddNat []))) (APP (APP (APP (CONST OfNat.ofNat [0]) (CONST Nat [])) (LIT 0)) (APP (CONST instOfNatNat []) (LIT 0)))) (APP (APP (APP (CONST OfNat.ofNat [0]) (CONST Nat [])) (LIT 1)) (APP (CONST instOfNatNat []) (LIT 1))))) (APP (APP (APP (CONST OfNat.ofNat [0]) (CONST Nat [])) (LIT 1)) (APP (CONST instOfNatNat []) (LIT 1)))))
```
Here is how it looks like after:
```
(APP (CONST Not []) (APP (APP (APP (CONST Eq [1]) (CONST Nat [])) (APP (APP (CONST Nat.add []) (LIT 0)) (LIT 1))) (LIT 1)))
```

Reply:
> Supporting `0 + 1 = 1` does not mean it's ok to reject `Nat.add 0 1 = 1`.

Indeed! However, the convention in Lean 4 seems to be to avoid concrete expressions which have typeclass versions. In fact, lots of builtins may break on these -- see [comment](https://github.com/leanprover/lean4/blob/7a3ee51d05b9ebbdbcc748a08be14b938df41b3e/src/Lean/Meta/DiscrTree.lean#L41). This is why I was worried about the unfolding pass.

Reply:
Hmm... I see your point. Using something like an `unfold` tactic for type class projections might solve both of our problems (staying true to the original goal and not having to deal with type classes at the SMT level).